### ISSUE TITLE
`pytest_generate_tests` conflicts with  `conftest.py` fixtures

### ISSUE THREAD

**Issue by vlcinsky** [pytest-dev/pytest#2726]
Dedicated repository for exhibiting the problem and with detailed description see https://github.com/vlcinsky/pytest_generate_tests-conflicts

From "summary" in repository README.md

### Duplicate values

As shown in `003_conftest_fixt1`, when `pytest_generate_tests` provided a value for test case
parameter and at the same time `conftest.py` provides a fixture with the same name, it wrongly
attempts to define a test case call also with the value from `conftest.py` fixture.

Expected behaviour is, that once a test function gets value for a parameter from
`pytest_generate_tests`, then further collection of values for this parameter shall stop.
Alternatively it shall resolve the duplication by giving the `pytest_generate_tests` parameter value
preference.

### Collection collects tests multiple times

As shown in `005_conftest_fixt2`, when `pytest_generate_tests` provides a value for a test case
parameter and at the same time in `conftest.py` there exist a fixture, which is dependent on another
one, it results in collecting the same test call (having the same combination of parameter values)
multiple times, this time it goes around `Duplicate values` by generating new test id using
parameter value from `conftest.py` fixture.

Expected behaviour is the same as described in "Duplicate values" above.

**Comment by nicoddemus** [comment]
@vlcinsky thanks a lot for all the effort you put into creating a repository and documenting each case, we really appreciate it.

**Comment by vlcinsky** [comment]
@nicoddemus This is my contribution to my beloved testing framework. Good quality issue description is rather easy comparing to reading pytest internals what I gave up being happy, others take care of it.

**Comment by RonnyPfannschmidt** [comment]
i believe this can be summarized as parametrize working unconditionally, so params and generate_tests are "mutually exclusive" 

both are implemented in terms of invoking metafunc.parametrize

fixing would to add support for either way would require overriding parametriting only if a parameter is not overridden

**Comment by fconil** [comment]
Thanks for you prompt reply.

In the [Python testing with pytest, 2nd ed](https://pragprog.com/titles/bopytest2/python-testing-with-pytest-second-edition/) book, the example uses `pytest_generate_tests` on a fixture that uses `params`.

You can look at chapter 5 [source code](https://media.pragprog.com/titles/bopytest2/code/bopytest2-code.zip), `test_fix_param.py` and `test_gen.py`.

Surprisingly the fixture code is not in `conftest.py` but in another test file.

If I put my fixture in another test file in the same folder than `test_hook.py`, it works.

**Comment by RonnyPfannschmidt** [comment]
indeed, it works if the individual hooks/configurations only apply to specific test files
in the book example each file has a self contained variant

if one moved all the implementations over to a conftest,
 one would start to see the conflicts as they would now apply to all files instead of individual files

**Comment by fconil** [comment]
I realized with this problem that I did not fully understand fixture loading / usage.

In the book example, `pytest_generate_tests` parametrize a fixture - `start_state` - that is not "defined" earlier. `start_state` does not come from `test_fix_param.py`  as I wrongly believed (thanks `--fixtures-per-test`).

This is the same for `stringinput` in the online documentation, https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-generate-tests

I note that I should not use `@pytest.fixture(params=...)` and `pytest_generate_tests` at the same time.

Thanks for your kind help !
